### PR TITLE
Support EKS

### PR DESCRIPTION
**Pull request by lawliet89**
# Breaking Changes

- Binary is now called `vault-k8s-helper`
- GKE users must now add a new `gke` positional argument before the Vault Path
- The payload returned from GKE now has the expiry key named `token_expiry` instead of the previous `expires_at_seconds`

EKS reference: https://github.com/kubernetes-sigs/aws-iam-authenticator#api-authorization-from-outside-a-cluster

### CODE REVIEW

File: src/aws.rs [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+    Ok(AwsCredentials::new(
+        creds.data.access_key,
+        creds.data.secret_key,
+        creds.data.security_token,
+        expiry,
+    ))
+}
+
+pub fn get_eks_token(
```
Comment (gyng) [inline src/aws.rs]: Move existing gke code to `gke.rs`?

Eg. `read_gcp_access_token`, `GcpAccessToken`

File: src/aws.rs [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+use serde::{Deserialize, Serialize};
+use vault::secrets::Aws;
+use vault::{self, Client};
+
+use crate::Error;
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
+pub struct EksCredential {
+    pub kind: &'static str,
+    #[serde(rename = "apiVersion")]
```
Comment (gyng) [inline src/aws.rs]: Why is this needed? AWS API? Maybe can add a link to the relevant docs here.

File: src/aws.rs [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+pub struct EksCredentialStatus {
+    pub token: String,
+}
+
+pub fn read_aws_credentials<S: AsRef<str>>(
+    client: &Client,
+    path: S,
+    request: &vault::secrets::aws::CredentialsRequest,
+) -> Result<AwsCredentials, Error> {
+    let path_parts: Vec<_> = path.as_ref().split('/').collect();
```
Comment (gyng) [inline src/aws.rs]: Why is this needed for AWS but not for GCP?

File: src/aws.rs [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    ))
+}
+
+pub fn get_eks_token(
+    credentials: &AwsCredentials,
+    cluster: &str,
+    region: Option<&str>,
+    expires_in: Option<&str>,
+) -> Result<EksCredential, Error> {
+    let region: Option<rusoto_core::region::Region> = match region {
```
Comment (gyng) [inline src/aws.rs]: https://doc.rust-lang.org/std/option/enum.Option.html#method.map?

File: src/aws.rs [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+        Some(r) => Some(r.parse()?),
+        None => None,
+    };
+    let expiry = match expires_in {
+        Some(expiry) => Some(std::time::Duration::from_secs(expiry.parse()?)),
+        None => None,
+    };
+    let headers = [("x-k8s-aws-id", cluster)].iter().cloned().collect();
```
Comment (gyng) [inline src/aws.rs]: Why is `iter().cloned().collect()` needed?

File: README.md [hunk 6]
```diff
@@ -1,28 +1,37 @@
...
 
+### AWS
+
+You have to configure Vault's
+[AWS Secrets Engine](https://www.vaultproject.io/docs/secrets/aws/index.html)
+with an appropriate role. Any role type should work.
+
 ```text
-Read Google Cloud Platform access token from Vault
+Read access tokens from Vault to authenticate with Kubernetes
```
[the remark this answers is not in the document]
Comment (lawliet89) [reply README.md]: No. This is a public repo.

File: src/aws.rs (hunk 2 above)
Comment (lawliet89) [reply src/aws.rs]: Can't find the docs. This is just what [aws-iam-authenticator](https://github.com/kubernetes-sigs/aws-iam-authenticator#) outputs.

File: src/aws.rs (hunk 3 above)
Comment (lawliet89) [reply src/aws.rs]: I haven't written a "wrapper" around the GCP secrets Engine. So it's just a straight up "get from the path" for GCP.

File: src/aws.rs (hunk 4 above)
Comment (lawliet89) [reply src/aws.rs]: Can't use `map`. `parse` can error and it's really messy to unwrap and deal with `Result`.

File: src/aws.rs (hunk 5 above)
Comment (lawliet89) [reply src/aws.rs]: Without `cloned`:

```
75 |     let headers = [("x-k8s-aws-id", cluster)].iter().collect();
   |                                                      ^^^^^^^ a collection of type `std::collections::HashMap<&str, &str>` cannot be built from `std::iter::Iterator<Item=&(&str, &str)>`
   |
   = help: the trait `std::iter::FromIterator<&(&str, &str)>` is not implemented for `std::collections::HashMap<&str, &str>`

```

`iter` turns the array into [slices](https://doc.rust-lang.org/std/primitive.slice.html) and you get a reference to the slice. I am just cloning the slice. A slice is already a "pointer" so this is just a cheap cloning of a pointer,

File: README.md [hunk 7]
```diff
@@ -33,6 +42,24 @@ FLAGS:
...
+        --eks-expiry <eks_expiry>
+            Specifies the Expiry duration in number of seconds for the Kubernetes Token.
+
+        --eks-region <eks_region>
+            Region of AWS to use. Defaults to the Global Endpoint
+
+        --eks-role-arn <eks_role_arn>
+            The ARN of the role to assume if the AWS Secrets Engine role is configured with multiple roles
+
+        --eks-ttl <eks_ttl>
```
[the remark this answers is not in the document]
Comment (lawliet89) [reply README.md]: Should be fixed.

File: README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (lawliet89) [reply README.md]: Done

File: src/aws.rs (hunk 1 above)
Comment (lawliet89) [reply src/aws.rs]: Done

**Comment by lawliet89** [comment]
```
   Compiling vault-k8s-helper v0.2.0 (/home/yongwen/work/open/vault-gke-helper)
    Finished dev [unoptimized + debuginfo] target(s) in 7.62s
     Running target/debug/deps/vault_k8s_helper-f335f522de6ea03e

running 4 tests
test tests::can_read_self_capabilities ... ok
test aws::tests::can_create_aws_token ... ok
test aws::tests::presigned_url_is_valid ... ok
test gcp::tests::can_read_gcp_secrets ... ok

test result: ok. 4 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
```
